router/http_test.go

```diff
@@ -869,6 +869,42 @@ func (s *S) runTestErrorAfterConnOnlyHitsOneBackend(c *C, upgrade bool) {
 	c.Assert(string(data), Equals, "Service Unavailable\n")
 }
 
+func (s *S) TestClientClosesConnectionBeforeResponse(c *C) {
+	gotReq := make(chan struct{})
+	backendDone := make(chan struct{})
+
+	h := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		close(gotReq)
+		defer close(backendDone)
+
+		cn := w.(http.CloseNotifier).CloseNotify()
+		select {
+		case <-cn:
+		case <-time.After(time.Second):
+			c.Fatal("expected reverse proxy to close the connection quickly")
+		}
+	})
+	srv := httptest.NewServer(h)
+	defer srv.Close()
+
+	l := s.newHTTPListener(c)
+	defer l.Close()
+
+	addHTTPRoute(c, l)
+	discoverdRegisterHTTP(c, l, srv.Listener.Addr().String())
+
+	req := newReq("http://"+l.Addr, "example.com")
+
+	go func() {
+		httpClient.Do(req)
+	}()
+
+	<-gotReq // wait for backend to get the request
+	httpClient.Transport.(*http.Transport).CancelRequest(req)
+
+	<-backendDone
```

Comment:
Should add selects with timeouts for both of these so that the tests don't hang if they break.